### PR TITLE
Add get_transaction_status to chain_plugin

### PR DESCRIPTION
**Pull request by oschwaldp-oci**
Adding a feature that allows querying the status of a transaction on the chain.

Support for the feature added in cleos via subcommand: `get transaction-status <transaction_id>`

Feature takes a transaction_id as input and outputs chain status and, if available, transaction status in the following format:

`{
  "state": "IN_BLOCK",
  "block_number": 90,
  "block_id": "0000005accfd59ba80a05380f60d51687406337b2aedd28b7daa33fdb8c16b5a",
  "block_timestamp": "2022-04-27T16:11:26.500",
  "expiration": "2022-04-27T16:11:56.000",
  "head_number": 186,
  "head_id": "000000bab27da51f76f483bb629b532510c22e2eb1acc632f5b37b421adecf63",
  "head_timestamp": "2022-04-27T16:12:14.500",
  "irreversible_number": 25,
  "irreversible_id": "0000001922118216bc16bf2c60d950598d80af3beca820eab751f7beecdb29e4",
  "irreversible_timestamp": "2022-04-27T16:10:54.000",
  "last_tracked_block_id": "000000129cee97f3e27312f0184d52d006a470f0e620553dfb4c5b4f3c856ab2"
}`

**Transaction Status:**

- If the transaction is unknown, failed (expired), in a block, rolled back, finalized, or limited in block (this would be for unknown trans ids that the controller knows about). This will be reported as `state` with values of: `UNKNOWN`, `FAILED`, `FORKED_OUT`, `LOCALLY_APPLIED`, `IN_BLOCK`, `IRREVERSIBLE`.
- The transaction’s block number, id, and timestamp, if they are applicable, will be reported as `block_number`, `block_id`, and `block_timestamp`. The transaction’s expiration date as `expiration`

**Chain Status:**

- The current head block number, id, and timestamp will be reported as `head_number`, `head_id`, and `head_timestamp`.
- The irreversible block number, id, and timestamp will be reported as `irreversible_number`, `irreversible_id`, and `irreversible_timestamp`.
- The _last_tracked_block_id, from the chain state object as `last_tracked_block_id`

### CODE REVIEW

File: plugins/chain_plugin/include/eosio/chain_plugin/chain_plugin.hpp [hunk 1]
```diff
@@ -126,6 +128,27 @@ class read_only {
...
+   struct get_transaction_status_results {
+      string                               state;
+      std::optional<uint32_t>              block_number;
+      std::optional<chain::block_id_type>  block_id;
+      std::optional<fc::time_point>        block_timestamp;
+      std::optional<fc::time_point>        expiration;
+      uint32_t                             head_number;
+      chain::block_id_type                 head_id;
+      fc::time_point                       head_timestamp;
+      uint32_t                             irreversible_number;
```
Comment (heifner) [inline plugins/chain_plugin/include/eosio/chain_plugin/chain_plugin.hpp:144]: Init `uint32_t`s

Comment (oschwaldp-oci) [reply plugins/chain_plugin/include/eosio/chain_plugin/chain_plugin.hpp]: Addressed in 16e0fd3848e7d23f79599796d0550b96d3cf6086

File: plugins/chain_api_plugin/chain_api_plugin.cpp [hunk 2]
```diff
@@ -161,6 +161,7 @@ void chain_api_plugin::plugin_startup() {
       CHAIN_RO_CALL(abi_bin_to_json, 200),
       CHAIN_RO_CALL(get_required_keys, 200),
       CHAIN_RO_CALL(get_transaction_id, 200),
+      CHAIN_RO_CALL(get_transaction_status, 200),
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply plugins/chain_api_plugin/chain_api_plugin.cpp]: Addressed in 16e0fd3848e7d23f79599796d0550b96d3cf6086

**Review by brianjohnson5972 — CHANGES_REQUESTED** [review]

File: plugins/chain_plugin/include/eosio/chain_plugin/chain_plugin.hpp [hunk 3]
```diff
@@ -126,6 +128,27 @@ class read_only {
    };
    get_info_results get_info(const get_info_params&) const;
 
+   struct get_transaction_status_params {
+      string                       id;
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply plugins/chain_plugin/include/eosio/chain_plugin/chain_plugin.hpp]: Fixed: 5c7c6eee7895b3b417000cac3b73be6e6f1ab350

File: tests/nodeos_run_test.py [hunk 4]
```diff
@@ -68,7 +68,8 @@
         cluster.killall(allInstances=killAll)
         cluster.cleanup()
         Print("Stand up cluster")
-        if cluster.launch(prodCount=prodCount, onlyBios=onlyBios, dontBootstrap=dontBootstrap) is False:
+        extraNodeosArgs=" --transaction-finality-status-max-storage-size-gb 1 "
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply tests/nodeos_run_test.py]: Fixed: 13eb845a5d317ecf0d18d8f8c40e205baf8bfdb6

File: plugins/chain_plugin/test/test_trx_finality_status_processing.cpp [hunk 5]
```diff
@@ -1004,7 +1005,9 @@ BOOST_AUTO_TEST_CASE(trx_finality_status_storage_reduction) { try {
 
    cs = status.get_chain_state();
    BOOST_CHECK(cs.head_id == b_12.bs->id);
+   BOOST_CHECK(cs.head_block_timestamp == b_12.bs->block->timestamp);
    BOOST_CHECK(cs.irr_id == eosio::chain::block_id_type{});
+   BOOST_CHECK(cs.irr_block_timestamp == eosio::chain::block_timestamp_type{});
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply plugins/chain_plugin/test/test_trx_finality_status_processing.cpp:1011]: Added test case:  c63c9abd62234d74afbb31200f59c0db08daddea

File: plugins/chain_plugin/chain_plugin.cpp [hunk 6]
```diff
@@ -1683,6 +1683,39 @@ read_only::get_info_results read_only::get_info(const read_only::get_info_params
...
+   auto input_id_length = param.id.size();
+   try
+   {
+      FC_ASSERT(input_id_length <= 64, "hex string is too long to represent an actual transaction id");
+      FC_ASSERT(input_id_length >= 8, "hex string representing transaction id should be at least 8 characters long to avoid excessive collisions");
+      input_id = transaction_id_type(param.id);
+   }
+   EOS_RETHROW_EXCEPTIONS(transaction_id_type_exception, "Invalid transaction ID: ${transaction_id}", ("transaction_id", param.id))
+
+   EOS_ASSERT(trx_finality_status_proc, unsupported_feature, "Transaction Status Interface not enabled.  To enable, configure nodeos with '--transaction-finality-status-max-storage-size-gb'.");
```
[the remark this answers is not in the document]
Comment (oschwaldp-oci) [reply plugins/chain_plugin/chain_plugin.cpp]: added:  65a43609c2d359e3059da90cccfc0210897c8f98

**Review by brianjohnson5972 — APPROVED** [review]

**Pull request merged**